### ISSUE TITLE
Sort order gets messed up when coupled to a redux flow

### ISSUE THREAD

**Issue by nervetattoo** [StreakYC/react-draggable-list#3]
First, thanks for the work on this library. React-draggable-list does all we need, except for the fact that we don't really get it to work alongside redux and redux-form.

The problem lies in the fact that react-draggable-list actually reorders the incoming `list` in `componentWillReceiveProps` when it's re-rendered after a drag has ended. What we do is something like:

``` js
<DraggableList
  ...
  list={listFromReduxStore}
  onMoveEnd={(list,item,oldIndex, newIndex) => {
    dispatchReduxResort('mylist', oldIndex, newIndex)
}} />
```

This _dispatch_ will in turn cause the redux store to build a new `listFromReduxStore` that is correctly sorted. When this list is passed to `react-draggable-list` it will get resorted in `componentWillReceiveProps` before the UI re-renders in a faulty state because the items were moved around based on their indexes, twice.

We're happy to contribute to a fix, but I'm not sure how you'd like to approach it.
I haven't dived deeply into the source, but maybe moving the splicing out from `componentWillReceiveProps` and into the move end handler is the way to go?

**Comment by AgentME** [comment]
I'm not sure I understand the issue. Doesn't `componentWillReceiveProps` only change the list if it's during a drag and the currently-dragged item is moved in the list? How does that reordering affect other things? Can you build a small example (based on example/Example.js if that makes it easiest) that shows the issue?

**Comment by nervetattoo** [comment]
I will see if I can create a small example tomorrow!

**Comment by nervetattoo** [comment]
Alright. I've built [this simple example that shows the error](https://nervetattoo.github.io/react-draggable-list-bug-redux/build/).
[And here's the source code](https://github.com/nervetattoo/react-draggable-list-bug-redux)

**Comment by AgentME** [comment]
Oh, so you're using the oldIndex and newIndex parameters to onMoveEnd and ignoring the newList parameter. The oldIndex/newIndex parameters have been pretty neglected and apparently don't accurately describe the move. We'll fix that.

At the current time, things will work if you ignore the oldIndex/newIndex parameters and use the newList argument, though I can see how it doesn't make for as nice of a redux action.

**Comment by nervetattoo** [comment]
It is some sort of timing issue. The component will first be re-rendered due to redux passing in a new list to react-draggable-list, and then `componentWillReceiveProps` excepts to get the *last* version of the list, and then it reorders itself, leaving the local state of the draggable-list out of sync with the redux state.
If you move https://github.com/StreakYC/react-draggable-list/blob/master/src/index.js#L112 to happen before calling `onMoveEnd` it will work because it will first reorder locally and then get the new state from redux and subsequentially re-render.

I can think of a few solutions on top of my head:
* Allow disabling componentWillReceiveProps
* Use the return value of `onMoveEnd` to dictate the new state

One thing _we_ could do, ofc, is to extend the class and override componentWillReceiveProps.

**Comment by AgentME** [comment]
I figured it out. The issue is here: https://github.com/nervetattoo/react-draggable-list-bug-redux/blob/0c6dbcf34e2f7bdc7b35c278b20b1b2b40360870/src/App.js#L41. You're setting the key to the item's current index. React-draggable-list (well, and React itself) identify the item by the key. When you move an item up the list, you're changing the keys of everything, including the item that was dragged, which trips react-draggable-list's special handling of the currently-dragged item being moved in the list. If you change the line to this, then it works, and also React knows it can move the elements instead of updating all of the elements to have new names:

```js
    const planets = this.props.planets.map(name => ({name,key: name}))
```

**Comment by nervetattoo** [comment]
That fixes the contrived example; However in our real codebase we already did this, with no luck. We do have some other components in the hierarchy that could be related to the issue, so I will see if I can expand upon my reproduction so it fails with proper keys set as well.
The real code is a complex, sortable form using redux-form, and nested react-draggable-list components, so theres more nuance to it ...

**Comment by AgentME** [comment]
>If you move https://github.com/StreakYC/react-draggable-list/blob/master/src/index.js#L112 to happen before calling onMoveEnd it will work because it will first reorder locally and then get the new state from redux and subsequentially re-render.

setState calls are asynchronous and won't be applied by the time componentWillReceiveProps is called. It's not available until componentWillUpdate, but then we can't call setState again during that step. We could remove state.list, and do the fix-up (based on state.lastDrag.startListKeys) in render and everywhere else that uses the list, which would fix us using the outdated version of state.drag immediately after onMoveEnd, but that's a bigger change. I'm currently leaning toward accepting that moving (or changing the key of; same thing to react and react-draggable-list) the dragged item in response to onMoveEnd results in undefined behavior, but anything else causing trouble here would be a good reason to make changes. Let me know if you can reproduce the issue with proper keying.

**Comment by nervetattoo** [comment]
Do you have any intuitive ideas about how we can forward in fixing this? We need to find a way to move forward, and I'd be happy to take a stab at a PR, but in that case it would be nice with some pointers in the direction you think it should go.

**Comment by skolmer** [comment]
I use this library in combination with redux-form and it's working great. 

The problem here is that the key of your list item component has to be the same even if the index of the item changes.

In your example posted above you are using the (reordered) index from the current redux form state (`fields`) to select a `key` from [`initialValues`](https://github.com/nervetattoo/react-draggable-list-bug-redux/blob/master/src/App.js#L57). Instead you should use `formValueSelector` to get the current state of your (reordered) list from the redux-form state and not use the list provided via `initialValues`. 
For more details please see the following issue https://github.com/erikras/redux-form/issues/1727#issue-175556288

**Comment by AgentME** [comment]
Your issue appears to be that you're not using the itemKey prop, similar to the previous issue. Is React not giving you a warning about the missing required prop?

**Comment by abrad45** [comment]
@AgentME thanks for the quick reply.

I am using the `itemKey` prop. I won't modify the previous example for posterity's sake, but to be more thorough, my component looks like this. `fieldset.id` is a uuid generated by https://www.npmjs.com/package/uuid

```js
const onDeleteStash = (index) => {
	fields.remove(index);
};

// Decorator: adds indices to each object in the FieldArray
// since DraggableList doesn't include them
const getListOfFieldsets = () => {
	return fields.getAll() && fields.getAll().length ?
		fields.getAll().map((fieldset, index) => Object.assign(fieldset, {index: index})) :
		[];
};

const onMoveEnd = (newList, movedStash, oldIndex, newIndex) => {
	fields.move(oldIndex, newIndex);
};

...

<DraggableList
	list={getListOfStashes()}
	itemKey="id"
	template={Parent}
	onMoveEnd={onMoveEnd}
	unsetZIndex
	commonProps={{
		onDeleteStash: onDeleteStash,
	}}
	/>
```

**Comment by AgentME** [comment]
Can you post a full example project that I could run that reproduces this issue? I haven't been able to reproduce this issue with proper itemKey usage.

**Comment by abrad45** [comment]
@AgentME this may be a little difficult, but I'll see if I can pull this off sometime soon.

**Comment by abrad45** [comment]
I set out to find a fix for this issue this morning and... I couldn't do it. I'm not sure why. The change I made? I set `itemKey` to the `index` I add in `getListOfFieldsets()`.

```js
// Decorator: adds indices to each object in the FieldArray
// since DraggableList doesn't include them
const getListOfFieldsets = () => {
	return fields.getAll() && fields.getAll().length ?
		fields.getAll().map((fieldset, index) => Object.assign(fieldset, {index: index})) :
		[];
};
...

<DraggableList
	list={getListOfFieldsets()}
	itemKey="index"
	template={Parent}
	onMoveEnd={onMoveEnd}
	unsetZIndex
	commonProps={{
		onDeleteStash: onDeleteStash,
	}}
	/>
```

I'm not entirely sure why this changes anything in a substantial way. If I understand what's going on correctly, `index` changes each time DraggableList is re-rendered because `getListOfFieldsets()` runs and rewrites each `index` value, whereas `id` is the `uuid` I was mentioning before and doesn't change. I'd expect `index` would be a _worse_ key than `id` but it appears to have resolved many of the issues I was seeing.

@AgentME does this behavior change make sense to you?

**Comment by AgentME** [comment]
Having itemKey set to index should cause the problem you mentioned earlier (it's exactly what caused this issue for someone above). If that's working now with itemKey=index, then I find that really surprising and more convincing that there is an issue in react-draggable-list. It's probably cancelling out the issue you ran into before, and will break again once that bug is fixed or if other changes are made to the project.

I still don't know what would cause this issue. I've used redux before, but I haven't found any reason why react-draggable-list would work differently with it, and I've failed at making my own reproduction cases. I would be able to investigate this more with a reproduction case.

**Comment by abrad45** [comment]
I'm glad we drew the same conclusions here. The issue is that the "reproducible case" is my project at work I cannot post code from, and the case that works is the demo I was able to post for you. By changing this value from `id` to `index` in my project at work, I've "fixed" (masked, as you propose) the problem. 🙃

**Comment by Macil** [comment]
I've just released v4.0 of react-draggable-list which I believe will solve this issue. I never managed to reproduce the issue, but I think I know what code was responsible, and I ripped that code out in the process of updating react-draggable-list to not use deprecated React features. DraggableList no longer keeps a separate copy of the list in state, which simplifies some things and should avoid this issue.
